src/main/java/com/limechain/trie/structure/decoded/node/DecodedNode.java

```diff
@@ -228,4 +239,218 @@ public byte[] calculateMerkleValue(Function<byte[], byte[]> hashFunction, boolea
+            if (pkLen > Integer.MAX_VALUE - addedValue) {
+                throw new NodeDecodingException("Partial key length overflow");
+            }
+
+            pkLen += addedValue;
+        }
+        return pkLen;
+    }
+
+    /**
+     * Extracts the partial key from a node in a trie structure using a ScaleCodecReader.
+     *
+     * @param reader The ScaleCodecReader used to read the encoded node data.
+     * @param pkLen  The length of the partial key in nibbles.
+     * @return byte[] The byte array containing the partial key.
+     * @throws NodeDecodingException If there are no more bytes to read, or if the partial key has invalid padding.
+     */
+    private static byte[] decodePartialKey(ScaleCodecReader reader, int pkLen) {
+        if (!reader.hasNext()) {
+            throw new NodeDecodingException("Node value cannot be null for partial key");
+        }
+
+        // Calculate the length of the partial key in bytes
+        int pkLenBytes = pkLen == 0 ? 0 : 1 + ((pkLen - 1) / 2);
+        byte[] partialKey = reader.readByteArray(pkLenBytes);
+
+        if ((pkLen % 2 == 1) && ((partialKey[0] & 0xf0) != 0)) {
+            throw new NodeDecodingException("Invalid Partial key padding");
+        }
+
+        return partialKey;
+    }
+
+    /**
+     * Extracts the children bitmap from a node in a trie structure.
+     *
+     * @param reader      The ScaleCodecReader used to read the encoded node data.
+     * @param hasChildren A boolean indicating whether the node has children.
+     * @return int The children bitmap as an integer. If the node has no children, returns 0.
+     * @throws NodeDecodingException If there are no more bytes to read, or if the children bitmap is zero.
+     */
+    private static int decodeChildrenBitmap(ScaleCodecReader reader, boolean hasChildren) {
+        if (!reader.hasNext()) {
+            throw new NodeDecodingException("Node value cannot be null for children bitmap");
+        }
+
+        if (!hasChildren) {
+            return 0;
+        }
+        byte[] bitmap;
+        final int BITMAP_DATA_LENGTH = 2;
+        bitmap = reader.readByteArray(BITMAP_DATA_LENGTH);
+        int childrenBitmap = ((bitmap[0] & 0xFF) | (bitmap[1] & 0xFF) << 8);
+
+        if (childrenBitmap == 0) {
+            throw new NodeDecodingException("Zero children bitmap");
+        }
+        return childrenBitmap;
+    }
+
+    /**
+     * Extracts the storage value from a node in a trie structure.
+     *
+     * @param reader             The ScaleCodecReader used to read the encoded node data.
+     * @param storageValueHashed A Boolean indicating whether the storage value is hashed. Can be null, true, or false.
+     * @return byte[] The storage value as a byte array, or null if there is no storage value.
+     * @throws NodeDecodingException If there are no more bytes to read from the reader.
+     */
+    private static byte[] decodeStorageValue(ScaleCodecReader reader, Boolean storageValueHashed) {
```

Comment:
Again, for intentionality (although mentioned in the Javadoc, but could be clarified a bit more):

```suggestion
    private static byte[] decodeStorageValue(ScaleCodecReader reader, @Nullable Boolean storageValueHashed)
```